python/singa/tensor.py

```diff
@@ -940,6 +940,309 @@ def mult(A, B, C=None, alpha=1.0, beta=0.0):
+    #to make the transpose match to its rank
+    transpose = sorted(range(len(transpose)), key = transpose.__getitem__)
+
+    return tensormult(A,B,sums,multiplies).transpose(transpose)
+
+
+def tensormult(A, B, sum=None, multiply=None):
+    '''
+    Args:
+        A (Singa.Tensor): The first input tensor from einsum.
+        B (Singa.Tensor): The second input tensor from einsum.
+        sum (list[list[int]]):
+            The axis to do the normal tensordot calculation
+
+        multiply (list[list[int]]):
+            The axis to multiply
+
+    Return: (Singa.Tensor)
+        The output of the tensormult calculation
+
+    fix some error of source code in the nils-werner/sparse
+    '''
+    if sum is None:
+        sum = [[], []]
+    else:
+        sum = list(sum)
+
+    if multiply is None:
+        multiply = [[], []]
+    else:
+        multiply = list(multiply)
+    # For each multiply[0] we are adding one axis, thus we need to increment
+    # all following items by one: (0, 1, 2) -> (0, 2, 4)
+    idx = multipliessort(multiply[0])
+
+    post_multiply = multiply[0]
+    for i, v in enumerate(idx):
+        post_multiply[v] += i
+
+
+#    for i in post_multiply:
+#        A = diag(A,i)
+    for i in sorted(post_multiply):
+
+        A = diag(A,i)
+        sum[0] = [x+1 if x >= i else x for x in sum[0]]
+
+    sum[0] += post_multiply
+    sum[1] += multiply[1]
+
+    return tensordot(A, B, axes=sum)
+
+def multipliessort(multiplies):
+    '''
+    Returns the indices that would sort an array
+    Args:
+        multiplies(list[int]):
+            the list of the input multiplies
+
+    Return(list[int]):
+        the indices list
+    '''
+
+    if multiplies is None:
+        multiplies = []
+    else:
+        multiplies = list(multiplies)
+
+    sort_multiplies = sorted(enumerate(multiplies), key=lambda x: x[1])
+    idx = [x[0] for x in sort_multiplies]
+    return idx
+
+def diag(A,axis=-1):
+    '''
+    Return the matrix that is diagonalize along the axis
+    Args:
+        A (Singa.Tensor):the input tensor
+        axis (int):
+            the axis which the matrix use to diagonalization
+    Return(Singa.Tensor):
+        the tensor which has been diagonalize along the given axis
+
+
+    It is like using numpy.diag() in the 1D condition, but this function can be used to do high dimension
+    matrix diagonalization
+
+    '''
+    # to get the shape of the diagonalize matirx
+    axis_diag = list(range(out.ndim))[axis]
+    shape = list(A.shape)
+    shape.insert(axis_diag,shape[axis_diag])
+
+    npA = to_numpy(A)
+    npA_diag = np.zeros(product(shape))
+    index = np.argwhere(npA != 0)
+    for i in index:
+        i_diag = list(i)
+        i_diag.insert(axis, i[axis])
+        index_npA = [[i[x]] for x in range(len(i))]
+        index_npA_diag = [[i_diag[x]] for x in range(len(i_diag))]
+        npA_diag[index_npA_diag] = npA[index_npA]
+    A_diag = from_numpy(npA_diag)
+    return A_diag
+
+def einsum2(A,B,ops):
```

Comment:
the previous implementation can be removed?
add  a TODO list to include the numpy functions to be replaced by singa functions.

Reply:
the einsum2 could realize the what the einsum do.  I will remove the previous einsum and add the TODO list.